bluefog/common/mpi_controller.cc

```diff
@@ -291,24 +296,104 @@ void MPIController::NeighborAllreduce(TensorTableEntry& entry) {
   // We need to explicitly set the device here.
   with_device device_guard(entry.device);
 
+  bool is_topo_check_fail = false;
+  if (!entry.send_neighbors->empty()) {
+    timeline_ptr->ActivityStart(entry.tensor_name, "NEGOTIATION");
+    int size = mpi_ctx_.size_;
+    int nsend = entry.send_neighbors->size();
+    int nrecv = entry.recv_neighbors->size();
+    bool* send_check_buf = new bool[2*size];
+    std::fill_n(send_check_buf, 2*size, false);
+    bool* recv_check_buf = new bool[2*size*size];
+    for (int send_rank : *(entry.send_neighbors))
+        send_check_buf[send_rank] = true;
+    for (int recv_rank : *(entry.recv_neighbors))
+        send_check_buf[mpi_ctx_.size_+recv_rank] = true;
+    int ret_code = MPI_Allgather(send_check_buf, size*2, MPI_C_BOOL,
+                                 recv_check_buf, size*2, MPI_C_BOOL,
+                                 mpi_ctx_.GetMPICommunicator(Communicator::GLOBAL));
+    if (ret_code != MPI_SUCCESS) {
+        throw std::runtime_error(
+            "MPI_Allgather (for dynamic neighbor_allreduce negotiation) failed, see MPI output "
+            "for details.");
+    }
+    auto get_index = [size](int i, int j) -> int { return 2*size*i+j; };
+    for (int i = 0; i < size; ++i) {
```

Comment:
I think the check is equivalent to 
Recv = [2\*sizes\*size] => two  boolean size*size matrix. One for recieving pattern and another for sending pattern. The rule is the transpose of recieving matrix should be the same for sending one, right?